accord-core/src/main/java/accord/impl/AbstractFetchCoordinator.java

```diff
@@ -0,0 +1,288 @@
+                    {
+                        default: throw new AssertionError("Unhandled enum: " + reply);
+                        case Invalid:
+                        case Redundant:
+                        case NotCommitted:
+                            throw new AssertionError(String.format("Unexpected reply: %s", reply));
+                        case Error:
+                            // TODO (required): ensure errors are propagated to coordinators and can be logged
+                    }
+                    return;
+                }
+
+                FetchResponse ok = (FetchResponse) reply;
+                Ranges received;
+                if (ok.unavailable != null)
+                {
+                    unavailable(to, ok.unavailable);
+                    if (ok.data == null)
+                    {
+                        inflight.remove(key).cancel();
+                        return;
+                    }
+                    received = ranges.difference(ok.unavailable);
+                }
+                else
+                {
+                    received = ranges;
+                }
+
+                // TODO (now): make sure it works if invoked in either order
+                inflight.remove(key).started(ok.maxApplied);
+                onReadOk(to, commandStore, ok.data, received);
+                // received must be invoked after submitting the persistence future, as it triggers onDone
+                // which creates a ReducingFuture over {@code persisting}
+            }
+
+            @Override
+            public void onFailure(Node.Id from, Throwable failure)
+            {
+                inflight.remove(key).cancel();
+                fail(from, failure);
+            }
+
+            @Override
+            public void onCallbackFailure(Node.Id from, Throwable failure)
+            {
+                // TODO (soon)
+                logger.error("Fetch coordination failure from " + from, failure);
+            }
+        });
+    }
+
+    @Override
+    protected synchronized void success(Node.Id to, Ranges ranges)
+    {
+        super.success(to, ranges);
+    }
+
+    @Override
+    protected synchronized void fail(Node.Id to, Ranges ranges, Throwable failure)
+    {
+        super.fail(to, ranges, failure);
+    }
+
+    public FetchResult result()
+    {
+        return result;
+    }
+
+    @Override
+    protected void onDone(Ranges success, Throwable failure)
+    {
+        if (success.isEmpty()) result.setFailure(failure);
+        else if (persisting.isEmpty()) result.setSuccess(Ranges.EMPTY);
+        else AsyncChains.reduce(persisting, (a, b) -> null)
+                        .begin((s, f) -> {
+                            if (f == null) result.setSuccess(ranges);
+                            else result.setFailure(f);
+                        });
+    }
+
+    @Override
+    public void start()
+    {
+        super.start();
+    }
+
+    void abort(Ranges abort)
```

Comment:
maybe make `abstract` for now?  `ListFetchCoordinator` is the only usage atm and could no-op it for now

Reply:
I'd rather leave it since it has a todo and is meant to be implemented later